### PR TITLE
Optimized TRT-LLM template truss

### PR DESCRIPTION
**Pull request by aspctu**
## Overview

This PR updates our base TRT-LLM template. Our current Truss is CPU bound, apparent at high loads where CPU hits 100% utilization. The previous fix was to start multiple instances of the Truss server (workers) but this doesn't tackle the root issue. This PR removes the actual bottleneck by leveraging an async GRPC client.

The main changes are:

1. Using async `predict` method instead of threading. 
2. Using an async GRPC client to stream responses
3. Creating one global GRPC stream for all requests 

## Benchmarks

We benchmark 4 variations of the truss:
- Standard synchronous truss with 1 worker
- Standard synchronous truss with 2 workers 
- Asynchronous truss with 1 worker
- Asynchronous truss with 2 workers

All benchmarks run with `k6`, 240 virtual users within a 60 second duration. The results are shown below. TLDR: Async truss is no longer CPU bound (as made apparent by seeing no benefit increasing num workers to 2, and by monitoring `top`. **The single worker async truss is more than 2x faster than the single worker sync truss and 28% faster than the 2 worker sync truss.** Finally, monitoring GPU utilization, we see more steady GPU utilization with the async truss vs. spiky utilization patterns of the sync truss at high load. 

### Standard synchronous truss with 1 worker
```
checks.........................: 100.00% ✓ 998       ✗ 0    
data_received..................: 593 kB  19 kB/s
data_sent......................: 218 kB  6.9 kB/s
http_req_blocked...............: avg=141.53µs min=1.97µs   med=4.11µs   max=18.08ms p(90)=283.52µs p(95)=477.54µs
http_req_connecting............: avg=49.56µs  min=0s       med=0s       max=2.45ms  p(90)=121.4µs  p(95)=255.35µs
http_req_duration..............: avg=7.46s    min=1.74s    med=7.75s    max=10.24s  p(90)=8.28s    p(95)=8.47s   
 { expected_response:true }...: avg=7.46s    min=1.74s    med=7.75s    max=10.24s  p(90)=8.28s    p(95)=8.47s   
http_req_failed................: 0.00%   ✓ 0         ✗ 998  
http_req_receiving.............: avg=6.44s    min=1.56s    med=6.66s    max=8.69s   p(90)=7.44s    p(95)=7.57s   
http_req_sending...............: avg=109.46µs min=9.53µs   med=21.54µs  max=4.95ms  p(90)=180.64µs p(95)=422.23µs
http_req_tls_handshaking.......: avg=0s       min=0s       med=0s       max=0s      p(90)=0s       p(95)=0s      
http_req_waiting...............: avg=1.02s    min=154.15ms med=999.23ms max=1.93s   p(90)=1.71s    p(95)=1.87s   
http_reqs......................: 998     31.704068/s
iteration_duration.............: avg=7.46s    min=1.74s    med=7.75s    max=10.24s  p(90)=8.28s    p(95)=8.47s   
iterations.....................: 998     31.704068/s
vus............................: 154     min=154     max=240
vus_max........................: 240     min=240     max=240
```

### Standard synchronous truss with 2 workers
```
checks.........................: 100.00% ✓ 3327      ✗ 0    
data_received..................: 2.0 MB  32 kB/s
data_sent......................: 725 kB  12 kB/s
http_req_blocked...............: avg=283.1µs  min=1.38µs  med=4.43µs   max=18.31ms  p(90)=7.75µs   p(95)=477.31µs
http_req_connecting............: avg=240.43µs min=0s      med=0s       max=16.03ms  p(90)=0s       p(95)=217.96µs
http_req_duration..............: avg=4.39s    min=1.39s   med=4.41s    max=7.93s    p(90)=4.91s    p(95)=5.07s   
 { expected_response:true }...: avg=4.39s    min=1.39s   med=4.41s    max=7.93s    p(90)=4.91s    p(95)=5.07s   
http_req_failed................: 0.00%   ✓ 0         ✗ 3327 
http_req_receiving.............: avg=4.17s    min=1.27s   med=4.24s    max=7.1s     p(90)=4.62s    p(95)=4.77s   
http_req_sending...............: avg=143.37µs min=7.41µs  med=26.4µs   max=10.7ms   p(90)=43.27µs  p(95)=245.88µs
http_req_tls_handshaking.......: avg=0s       min=0s        med=0s       max=0s       p(90)=0s       p(95)=0s      
http_req_waiting...............: avg=222.71ms min=51.49ms med=160.49ms max=838.34ms p(90)=449.12ms p(95)=640.56ms
http_reqs......................: 3327    53.883226/s
iteration_duration.............: avg=4.4s     min=1.39s   med=4.41s    max=7.93s    p(90)=4.91s    p(95)=5.07s   
iterations.....................: 3327    53.883226/s
vus............................: 136     min=136     max=240
vus_max........................: 240     min=240     max=240
```

### Asynchronous truss with 1 worker
```
checks.........................: 100.00% ✓ 4364      ✗ 0    
data_received..................: 2.6 MB  41 kB/s
data_sent......................: 953 kB  15 kB/s
http_req_blocked...............: avg=30.21µs min=1.5µs    med=3.62µs  max=4.41ms   p(90)=5.54µs  p(95)=121.12µs
http_req_connecting............: avg=18.45µs min=0s       med=0s      max=4.12ms   p(90)=0s      p(95)=73.13µs 
http_req_duration..............: avg=3.39s   min=929.04ms med=3.46s   max=4.06s    p(90)=3.65s   p(95)=3.7s    
 { expected_response:true }...: avg=3.39s   min=929.04ms med=3.46s   max=4.06s    p(90)=3.65s   p(95)=3.7s    
http_req_failed................: 0.00%   ✓ 0         ✗ 4364 
http_req_receiving.............: avg=3.35s   min=545.94ms med=3.44s   max=3.81s    p(90)=3.63s   p(95)=3.67s   
http_req_sending...............: avg=39.83µs min=9.23µs   med=19.27µs max=6.13ms   p(90)=34.39µs p(95)=50.88µs 
http_req_tls_handshaking.......: avg=0s      min=0s       med=0s      max=0s       p(90)=0s      p(95)=0s      
http_req_waiting...............: avg=40.26ms min=1.72ms   med=15.91ms max=413.32ms p(90)=38.89ms p(95)=378.08ms
http_reqs......................: 4364    68.926219/s
iteration_duration.............: avg=3.39s   min=929.68ms med=3.46s   max=4.06s    p(90)=3.65s   p(95)=3.7s    
iterations.....................: 4364    68.926219/s
vus............................: 37      min=37      max=240
vus_max........................: 240     min=240     max=240
```

### Asynchronous truss with 2 workers
```
 checks.........................: 100.00% ✓ 4277      ✗ 0    
 data_received..................: 2.5 MB  40 kB/s
 data_sent......................: 935 kB  15 kB/s
 http_req_blocked...............: avg=75.65µs min=1.8µs    med=3.57µs  max=10.78ms  p(90)=5.32µs  p(95)=181.81µs
 http_req_connecting............: avg=51.44µs min=0s       med=0s      max=4.81ms   p(90)=0s      p(95)=80.47µs 
 http_req_duration..............: avg=3.45s   min=953.53ms med=3.48s   max=4.22s    p(90)=3.85s   p(95)=3.99s   
   { expected_response:true }...: avg=3.45s   min=953.53ms med=3.48s   max=4.22s    p(90)=3.85s   p(95)=3.99s   
 http_req_failed................: 0.00%   ✓ 0         ✗ 4277 
 http_req_receiving.............: avg=3.41s   min=699.98ms med=3.45s   max=4.19s    p(90)=3.81s   p(95)=3.92s   
 http_req_sending...............: avg=73.55µs min=9.81µs   med=19.73µs max=6.61ms   p(90)=35.92µs p(95)=110.19µs
 http_req_tls_handshaking.......: avg=0s      min=0s       med=0s      max=0s       p(90)=0s      p(95)=0s      
 http_req_waiting...............: avg=40.41ms min=2.02ms   med=23.67ms max=265.86ms p(90)=70.84ms p(95)=246.82ms
 http_reqs......................: 4277    67.934707/s
 iteration_duration.............: avg=3.45s   min=954.24ms med=3.48s   max=4.22s    p(90)=3.85s   p(95)=3.99s   
 iterations.....................: 4277    67.934707/s
 vus............................: 101     min=101     max=240
 vus_max........................: 240     min=240     max=240
```

### CODE REVIEW

File: templates/trt-llm/packages/client.py [hunk 1]
```diff
@@ -122,34 +56,99 @@ def load_server_and_model(self, env: dict):
...
-        """Static method to yield predictions or errors based on input and a streaming user_data queue."""
+    def start_server(
+        self,
+        mpi: int = 1,
+        env: dict = {},
+    ):
+        """Triton Inference Server has different startup commands depending on
+        whether it is running in a TP=1 or TP>1 configuration. This function
+        starts the server with the appropriate command."""
+        def _build_server_start_command(
```
Comment (pankajroark) [inline templates/trt-llm/packages/client.py:67]: nit: Seems like this function can be easily moved outside and doesn't need to be an inner function.

File: templates/trt-llm/packages/client.py [hunk 2]
```diff
@@ -122,34 +56,99 @@ def load_server_and_model(self, env: dict):
...
+        ):
+            base_command = [
+                "tritonserver",
+                "--model-repository", str(self._model_repository_dir),
+                "--grpc-port", str(GRPC_SERVICE_PORT),
+                "--http-port", str(HTTP_SERVICE_PORT)
+            ]
+
+            if mpi == 1:
+                # return base_command + ["--log-verbose", "1"]
```
Comment (pankajroark) [inline templates/trt-llm/packages/client.py:79]: Maybe explain how the commented code could be used for getting more logs from triton server.

**Review by pankajroark — APPROVED** [review]
Great! Much simpler code!

**Review by pankajroark — APPROVED** [review]
Great! Much simpler code!

**Comment by amiruci** [comment]
super!!

**Comment by joostinyi** [comment]
Let's make a big note to remind ourselves to reinstate the generation pre commit hooks for other TRT-LLM trusses after this lands – whatever shape that ends up taking

**Review by joostinyi — APPROVED** [review]

**Pull request merged**